crates/types/src/entities/coins.rs

```diff
@@ -73,10 +73,10 @@ impl CoinType {
     }
 
     /// Returns the asset held by the coin.
-    pub fn asset_id(&self) -> &AssetId {
+    pub fn asset_id<'a>(&'a self, base_asset_id: &'a AssetId) -> &'a AssetId {
         match self {
             CoinType::Coin(coin) => &coin.asset_id,
-            CoinType::MessageCoin(_) => &AssetId::BASE,
+            CoinType::MessageCoin(_) => base_asset_id,
         }
```

Comment:
This feels a bit weird to me, shpuld we return `Option<AssetID>` instead, returning `None` for coins? Then the caller could do `.asset_id().unwrap_or(base_asset_id)`.

Reply:
I understand the motivation behind your suggestion, but I would argue that returning `Option` changes the semantics of the signature in way that is incorrect: a coin must always have an asset, in the same way that a coin must always have an amount. It is the unit in which the scalar quantity is measured. 

If we use `Option`, it would require introducing `.unwrap_or(base_asset_id)` as a ubiquitous idiom, but we can't really enforce that. 

I agree that passing the `base_asset_id` to the getter is a weird idiom itself. Unfortunately, this is the most straightforward approach, since we cannot store the asset ID on messages themselves - the serialization/ deserialization specification of messages excludes asset ID, so we have no guaranteed way of querying it from the message data. We need to feed it in externally.